hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisOverReplicationHandler.java

```diff
@@ -0,0 +1,247 @@
+            "with {} pending delete(s). Expected replica count is {}.",
+        containerInfo.containerID(),
+        replicaCount.getReplicas().size(), pendingDelete,
+        replicaCount.getReplicationFactor());
+
+    // get replicas that can be deleted, in sorted order
+    List<ContainerReplica> eligibleReplicas =
+        getEligibleReplicas(replicaCount, pendingOps);
+    if (eligibleReplicas.size() == 0) {
+      LOG.info("Did not find any replicas that are eligible to be deleted for" +
+          " container {}.", containerInfo);
+      return Collections.emptyMap();
+    }
+
+    return createCommands(containerInfo, eligibleReplicas, excess);
+  }
+
+  private boolean verifyOverReplication(
+      RatisContainerReplicaCount replicaCount) {
+    if (!replicaCount.isOverReplicated()) {
+      LOG.info("Container {} is actually not over-replicated any more.",
+          replicaCount.getContainer().containerID());
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Finds replicas that are eligible to be deleted, sorted to avoid
+   * potential data loss.
+   * @see
+   * <a href="https://issues.apache.org/jira/browse/HDDS-4589">HDDS-4589</a>
+   * @param replicaCount ContainerReplicaCount object for the container
+   * @param pendingOps Pending adds and deletes
+   * @return List of ContainerReplica sorted using
+   * {@link RatisOverReplicationHandler#sortReplicas(Collection)}
+   */
+  private List<ContainerReplica> getEligibleReplicas(
+      ContainerReplicaCount replicaCount, List<ContainerReplicaOp> pendingOps) {
+    // sort replicas so that they can be selected in a deterministic way
+    List<ContainerReplica> eligibleReplicas =
+        sortReplicas(replicaCount.getReplicas());
+
+    // retain one replica per unique origin datanode if the container is not
+    // closed
+    final Map<UUID, ContainerReplica> uniqueReplicas =
+        new LinkedHashMap<>();
+    if (replicaCount.getContainer().getState() !=
+        HddsProtos.LifeCycleState.CLOSED) {
+      eligibleReplicas.stream()
+          // get replicas with state that matches container state
+          .filter(r -> ReplicationManager.compareState(
+              replicaCount.getContainer().getState(),
+              r.getState()))
+          .forEach(r -> uniqueReplicas
+              .putIfAbsent(r.getOriginDatanodeId(), r));
+
+      // note that this preserves order of the List
+      eligibleReplicas.removeAll(uniqueReplicas.values());
+    }
+
+    Set<DatanodeDetails> pendingDeletion = new HashSet<>();
+    // collect the DNs that are going to have their container replica deleted
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        pendingDeletion.add(op.getTarget());
+      }
+    }
+
+    // replicas that are not on IN_SERVICE nodes or are already pending
+    // delete are not eligible
+    // TODO what about nodes that are not healthy?
```

Comment:
Legacy RM does not filter healthy nodes but I think we should.

Reply:
That is an interesting question. An IN_SERVICE node that is not healthy must be stale. If it was dead, the replicas should be removed from SCM for it, but perhaps there is a short race condition where that could not happen.

If you happen to have a DEAD or a STALE replica, we can assume the DEAD will go away very soon by itself. The STALE may well become DEAD and go away too.

So perhaps we should decrement the excess by any stale or dead we find, and not send a command for them, as they are going to go away naturally anyway, or come back to healthy.

Otherwise, we could schedule a delete on a healthy, and then the stale / dead goes away and we end up under-replicated.